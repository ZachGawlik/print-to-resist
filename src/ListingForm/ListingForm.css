form {
  text-align: left;
}

label {
  display: block;
}
