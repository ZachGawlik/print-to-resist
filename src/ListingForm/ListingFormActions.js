function checkStatus(response) {
  if (response.status >= 200 && response.status < 300) {
    return response;
  }
  const error = new Error(response.statusText);
  error.response = response;
  throw error;
}

export function submitForm(data) {
  fetch('http://localhost:3000/poster', {
    method: 'POST',
    body: data
  })
  .then(checkStatus)
  .then(response => {
    console.log('success');
  })
  .catch(err => {
    console.log(err);
  });
}
