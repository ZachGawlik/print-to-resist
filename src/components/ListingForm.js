import React, { PropTypes } from 'react';
import moment from 'moment';
import Dropzone from 'react-dropzone';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import { WithContext as ReactTags } from 'react-tag-input';
import LinkInput from './LinkInput';
import {
  COLOR_OPTIONS,
  PAPER_SIZES
} from '../constants';
import '../styles/ReactTags.css'

class ListingForm extends React.Component {
  static propTypes = {
    handleAddLink: PropTypes.func.isRequired,
    handleColorInput: PropTypes.func.isRequired,
    handleDateChange: PropTypes.func.isRequired,
    handleDeleteLink: PropTypes.func.isRequired,
    handleInputChange: PropTypes.func.isRequired,
    handlePosterChange: PropTypes.func.isRequired,
    handleTagAddition: PropTypes.func.isRequired,
    handleTagDelete: PropTypes.func.isRequired,
    handleThumbnailChange: PropTypes.func.isRequired,
    handleUrlChange: PropTypes.func.isRequired,
    listing: PropTypes.object.isRequired,
    postListing: PropTypes.func.isRequired
  }

  isValid = () => {
    const { listing } = this.props;
    return (
      !!listing.description &&
      !!listing.poster &&
      !!listing.title &&
      !!listing.thumbnail
    );
  }

  handleSubmit = (event) => {
    const { listing } = this.props;
    const data = new FormData();
    Object.keys(listing).forEach((field) => {
      if (field === 'tags') {
        listing.tags.forEach(tag => {
          data.append(field, tag.id);
        })
      } else if (field === 'links') {
        listing.links.forEach(link => {
          data.append(field, link)
        });
      } else {
        data.append(field, listing[field]);
      }
    })

    if (this.isValid()) {
      this.props.postListing(data);
    }
    event.preventDefault();
  }

  render() {
    const { listing } = this.props;
    return (
      <form onSubmit={this.handleSubmit}>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="thumbnail-input"
          >
            Thumbnail:
            <span className="ListingForm__input-info">(must be less than 10MB)</span>
          </label>
          <div>
            {listing.thumbnail &&
              <img
                alt="thumbnail"
                className="ListingForm__preview"
                src={listing.thumbnail.preview}
              />
            }
            <Dropzone
              accept="image/jpeg, image/png"
              className="ListingForm__dropzone"
              inputProps={{id: 'thumbnail-input'}}
              multiple={false}
              onDropAccepted={this.props.handleThumbnailChange}
            >
              Add your thumbnail
            </Dropzone>
          </div>
        </div>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="title-input"
          >
            Title:
          </label>
          <input
            id="title-input"
            name="title"
            type="text"
            onChange={this.props.handleInputChange}
          />
        </div>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="poster-input"
          >
            Poster:
            <span className="ListingForm__input-info">(must be less than 10MB)</span>
          </label>
          {listing.poster &&
            listing.poster.name}
          <Dropzone
            accept="image/jpeg, image/png, application/pdf"
            className="ListingForm__dropzone"
            disablePreview={true}
            inputProps={{id: 'poster-input'}}
            multiple={false}
            onDropAccepted={this.props.handlePosterChange}
          >
            Add your poster
          </Dropzone>
        </div>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="description"
          >
            Description
          </label>
          <textarea
            id="description"
            name="description"
            onChange={this.props.handleInputChange}
            value={listing.description}
          />
        </div>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="instruction"
          >
            Special instruction:
            <span className="ListingForm__input-info">Where/how should this be spread?</span>
          </label>
          <textarea
            id="instruction"
            name="instruction"
            onChange={this.props.handleInputChange}
            value={listing.instruction}
          />
        </div>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="tags"
          >
            Tags
            <span className="ListingForm__input-info">(up to 3)</span>
          </label>
          <ReactTags
            autofocus={false}
            classNames={{
              tagInput: listing.tags.length >= 3 ? 'display-none' : 'ReactTags__tagInput'
            }}
            handleAddition={this.props.handleTagAddition}
            handleDelete={this.props.handleTagDelete}
            id="tags"
            minQueryLength={1}
            name="tags"
            tags={listing.tags}
          />
        </div>
        <div className="ListingForm__field">
          <label className="ListingForm__field-label">
            Links
          </label>
          {listing.links.map((link, index) =>
            <LinkInput
              key={index}
              index={index}
              onDeleteLink={this.props.handleDeleteLink}
              onUrlChange={this.props.handleUrlChange}
              url={link}
            />
          )}
          <button
            onClick={this.props.handleAddLink}
            type="button"
          >
            Add +
          </button>
        </div>
        <div className="ListingForm__field">
          <label
            className="ListingForm__field-label"
            htmlFor="deadline"
          >
            Deadline
          </label>
          <DatePicker
            isClearable={true}
            minDate={moment()}
            maxDate={moment().add(1, "year")}
            onChange={this.props.handleDateChange}
            placeholderText="No deadline selected"
            selected={listing.deadline}
          />
        </div>
        <div className="ListingForm__field">
          <p className="ListingForm__field-label">Print in:</p>
          <label className="ListingForm__radio-label">
            <input
              type="radio"
              name="isColor"
              value={COLOR_OPTIONS.bw}
              checked={!listing.isColor}
              onChange={this.props.handleColorInput}
            />
            B/W
          </label>
          <label className="ListingForm__radio-label">
            <input
              type="radio"
              name="isColor"
              value={COLOR_OPTIONS.color}
              checked={!!listing.isColor}
              onChange={this.props.handleColorInput}
            />
            <span className="colorful">Color</span>
          </label>
        </div>
        <div className="ListingForm__field">
          <p className="ListingForm__field-label">Page size</p>
          <label className="ListingForm__radio-label">
            <input
              type="radio"
              name="paperSize"
              value={PAPER_SIZES.letter}
              checked={listing.paperSize === PAPER_SIZES.letter}
              onChange={this.props.handleInputChange}
            />Letter
          </label>
          <label className="ListingForm__radio-label">
            <input
              type="radio"
              name="paperSize"
              value={PAPER_SIZES.tabloid}
              checked={listing.paperSize === PAPER_SIZES.tabloid}
              onChange={this.props.handleInputChange}
            />Tabloid (17 x 11)
          </label>
        </div>
        <input type="submit" value="Submit" disabled={!this.isValid()} />
      </form>
    );
  }
}

export default ListingForm;
