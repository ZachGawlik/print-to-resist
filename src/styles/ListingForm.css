.ListingForm {
  margin: 0 auto;
  max-width: 600px;
}

.ListingForm__preview {
  max-width: 200px;
  max-height: 200px;
}

.ListingForm__field {
  margin-bottom: 2em;
}

.ListingForm__field-label {
  margin-bottom: .25em;
  display: block;
  font-weight: bold;
}

.ListingForm__radio-label {
  display: block;
  font-weight: normal
}

.ListingForm__disclaimer {
  color: grey;
  margin-bottom: 2em;
}

.ListingForm__input-info {
  margin-left: .5em;
  color: grey;
  font-weight: normal;
}

.ListingForm__dropzone {
  width: 200px;
  height: 200px;
  border: 1px solid rgba(0,0,0,.15);
  border-radius: 5px;
  padding: 1em;
  cursor: pointer;
}
