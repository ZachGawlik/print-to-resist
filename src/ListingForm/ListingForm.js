import React from 'react';
import _ from 'lodash';
import moment from 'moment';
import Dropzone from 'react-dropzone';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import './ListingForm.css';
import { submitForm } from './ListingFormActions';

const PAPER_SIZES = {
  letter: 'letter',
  tabloid: 'tabloid'
};

class ListingForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      deadline: null,
      description: '',
      instructions: '',
      isColor: false,
      paperSize: PAPER_SIZES.letter,
      poster: null,
      thumbnail: null,
      testtest: null
    };
  }

  handleDateChange = (newDate) => {
    this.setState({
      deadline: newDate
    });
  }

  handleInputChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  handleSubmit = (event) => {
    const data = new FormData();
    _.forEach(Object.keys(this.state), (field) => {
      data.append(field, this.state[field]);
    })

    submitForm(data);
    event.preventDefault();
  }

  handleThumbnailChange = (acceptedFiles) => {
    this.setState({
      thumbnail: acceptedFiles[0]
    });
  }

  handlePosterChange = (acceptedFiles) => {
    this.setState({
      poster: acceptedFiles[0]
    });
  }

  // TODO: validate required fields
  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Thumbnail:
        </label>
        <div>
          {this.state.thumbnail &&
            <img alt="thumbnail" src={this.state.thumbnail.preview} />
          }
          <Dropzone
            accept="image/jpeg, image/png, application/pdf"
            onDropAccepted={this.handleThumbnailChange}
          >
            Add your thumbnail
          </Dropzone>
        </div>
        <label htmlFor="title">
          Title:
        </label>
        <input
          name="title"
          type="text"
          onChange={this.handleInputChange}
        />
        <div>
          <label>
            Poster:
          </label>
          <Dropzone
            accept="image/jpeg, image/png, application/pdf"
            disablePreview={true}
            onDropAccepted={this.handlePosterChange}
          >
            Add your Poster.
          </Dropzone>
        </div>
        <label htmlFor="description">
          Description
        </label>
        <textarea
          name="description"
          onChange={this.handleInputChange}
          value={this.state.description}
        />
        <label htmlFor="instructions">
          Special instructions:
        </label>
        <textarea
          name="instructions"
          onChange={this.handleInputChange}
          value={this.state.instructions}
        />
        <div>
          <label>Deadline</label>
          <DatePicker
            isClearable={true}
            minDate={moment()}
            maxDate={moment().add(1, "year")}
            onChange={this.handleDateChange}
            placeholderText="deadline"
            selected={this.state.deadline}
          />
        </div>
        <div>
          <h3>Print in:</h3>
          <label>
            <input
              type="radio"
              name="isColor"
              value={false}
              checked={!this.state.isColor}
              onChange={this.handleInputChange}
            />Black and White
          </label>
          <label>
            <input
              type="radio"
              name="isColor"
              value={true}
              checked={this.state.isColor}
              onChange={this.handleInputChange}
            />Color
          </label>
        </div>
        <div>
          <h3>Page size</h3>
          <label>
            <input
              type="radio"
              name="paperSize"
              value="letter"
              checked={this.state.paperSize === PAPER_SIZES.letter}
              onChange={this.handleInputChange}
            />Letter
          </label>
          <label>
            <input
              type="radio"
              name="paperSize"
              value="tabloid"
              checked={this.state.paperSize === PAPER_SIZES.tabloid}
              onChange={this.handleInputChange}
            />Tabloid (17 x 11)
          </label>
        </div>
        <input type="submit" value="Submit" />
      </form>
    )
  }
}

export default ListingForm;
