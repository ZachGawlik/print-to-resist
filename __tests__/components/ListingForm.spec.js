import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import ListingForm from '../../src/components/ListingForm';

const blankListing = {
  deadline: null,
  description: '',
  instruction: '',
  isColor: false,
  links: [],
  paperSize: 'Letter',
  poster: null,
  tags: [],
  title: '',
  thumbnail: null
};

describe('ListingForm', () => {
  it('renders as expected', () => {
    Date.now = jest.fn(() => 1482363367071);
    const renderer = new ShallowRenderer();
    renderer.render(
      <ListingForm
        handleAddLink={jest.fn()}
        handleColorInput={jest.fn()}
        handleDateChange={jest.fn()}
        handleDeleteLink={jest.fn()}
        handleInputChange={jest.fn()}
        handlePosterChange={jest.fn()}
        handleTagAddition={jest.fn()}
        handleTagDelete={jest.fn()}
        handleThumbnailChange={jest.fn()}
        handleUrlChange={jest.fn()}
        listing={blankListing}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
