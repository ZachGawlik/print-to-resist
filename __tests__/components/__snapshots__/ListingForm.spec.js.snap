// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingForm renders as expected 1`] = `
<div>
  <h2>
    Submit a poster worth spreading
  </h2>
  <p
    className="ListingForm__disclaimer"
  >
    The submission will be reviewed to ensure the poster is accurate and relevant. If the poster pertains to a specific event, the organizer will be contacted to ensure that they agree to having the event be spread in this way.
  </p>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="thumbnail-input"
    >
      Thumbnail:
      <span
        className="ListingForm__input-info"
      >
        (preferably 300x300)
      </span>
    </label>
    <div>
      <Dropzone
        accept="image/jpeg, image/png"
        className="ListingForm__dropzone"
        disableClick={false}
        disablePreview={false}
        inputProps={
          Object {
            "id": "thumbnail-input",
          }
        }
        maxSize={Infinity}
        minSize={0}
        multiple={false}
        onDropAccepted={[Function]}
        preventDropOnDocument={true}
      >
        Add your thumbnail
      </Dropzone>
    </div>
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="title-input"
    >
      Title:
    </label>
    <input
      id="title-input"
      name="title"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="poster-input"
    >
      Poster:
      <span
        className="ListingForm__input-info"
      >
        (must be less than 10MB)
      </span>
    </label>
    <Dropzone
      accept="image/jpeg, image/png, application/pdf"
      className="ListingForm__dropzone"
      disableClick={false}
      disablePreview={true}
      inputProps={
        Object {
          "id": "poster-input",
        }
      }
      maxSize={Infinity}
      minSize={0}
      multiple={false}
      onDropAccepted={[Function]}
      preventDropOnDocument={true}
    >
      Add your poster
    </Dropzone>
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="description"
    >
      Description
    </label>
    <textarea
      id="description"
      name="description"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="instruction"
    >
      Special instruction:
      <span
        className="ListingForm__input-info"
      >
        Where/how should this be spread?
      </span>
    </label>
    <textarea
      id="instruction"
      name="instruction"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="tags"
    >
      Tags
      <span
        className="ListingForm__input-info"
      >
        (up to 3)
      </span>
    </label>
    <DragDropContext(ReactTags)
      autofocus={false}
      classNames={
        Object {
          "tagInput": "ReactTags__tagInput",
        }
      }
      handleAddition={[Function]}
      handleDelete={[Function]}
      id="tags"
      minQueryLength={1}
      name="tags"
      tags={Array []}
    />
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
    >
      Links
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Add +
    </button>
  </div>
  <div
    className="ListingForm__field"
  >
    <label
      className="ListingForm__field-label"
      htmlFor="deadline"
    >
      Deadline
      <span
        className="ListingForm__input-info"
      >
        (if applicable)
      </span>
    </label>
    <t
      dateFormat="L"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      isClearable={true}
      maxDate={"2017-12-21T23:36:07.071Z"}
      minDate={"2016-12-21T23:36:07.071Z"}
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      placeholderText="No deadline selected"
      popoverAttachment="top left"
      popoverTargetAttachment="bottom left"
      popoverTargetOffset="10px 0"
      selected={null}
      tetherConstraints={
        Array [
          Object {
            "attachment": "together",
            "to": "window",
          },
        ]
      }
      utcOffset={-300}
      withPortal={false}
    />
  </div>
  <div
    className="ListingForm__field"
  >
    <p
      className="ListingForm__field-label"
    >
      Print in:
    </p>
    <label
      className="ListingForm__radio-label"
    >
      <input
        checked={true}
        name="isColor"
        onChange={[Function]}
        type="radio"
        value="Black and White"
      />
      B/W
    </label>
    <label
      className="ListingForm__radio-label"
    >
      <input
        checked={false}
        name="isColor"
        onChange={[Function]}
        type="radio"
        value="Color"
      />
      <span
        className="colorful"
      >
        Color
      </span>
    </label>
  </div>
  <div
    className="ListingForm__field"
  >
    <p
      className="ListingForm__field-label"
    >
      Page size
    </p>
    <label
      className="ListingForm__radio-label"
    >
      <input
        checked={true}
        name="paperSize"
        onChange={[Function]}
        type="radio"
        value="Letter"
      />
      Letter
    </label>
    <label
      className="ListingForm__radio-label"
    >
      <input
        checked={false}
        name="paperSize"
        onChange={[Function]}
        type="radio"
        value="Tabloid"
      />
      Tabloid (17 x 11)
    </label>
  </div>
</div>
`;
